sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -468,7 +489,13 @@ case class DataSource(
       throw new AnalysisException("Cannot save interval data type into external storage.")
     }
 
-    providingClass.newInstance() match {
+    val resolvedRelation = providingClass.newInstance() match {
+      case relationToCheck: DataSourceValidator =>
```

Comment:
> In other cases, we need to ad-hoc check the duplication (e.g., JDBCRelation)

@cloud-fan How about this? Since we couldn't pass `df.schema` into the check in `resolveRelation` you suggested, so I put the check here for write paths. Actually, IMHO we need not have this datasource-specific check for read paths because each datasource implementation should provide a valid schema when inferring it in `FileFormat.inferSchema`, `JdbcUtils.getSchema`, ...

On the other hand, in write paths, I feel other datasource-specific checks would be better to be done in `DataSource`. For example;
```
scala> spark.range(1).selectExpr("rand()").write.save("path")
org.apache.spark.sql.AnalysisException: Attribute name "rand(1595701563628455153)" contains invalid character(s) among " ,;{}()\n\t=". Please use alias to rename it.;
  at org.apache.spark.sql.execution.datasources.parquet.ParquetSchemaConverter$.checkConversionRequirement(ParquetSchemaConverter.scala:581)
  at org.apache.spark.sql.execution.datasources.parquet.ParquetSchemaConverter$.checkFieldName(ParquetSchemaConverter.scala:567)
  at org.apache.spark.sql.execution.datasources.parquet.ParquetWriteSupport$$anonfun$setSchema$2.apply(ParquetWriteSupport.scala:446)
  at org.apache.spark.sql.execution.datasources.parquet.ParquetWriteSupport$$anonfun$setSchema$2.apply(ParquetWriteSupport.scala:446)
  at scala.collection.immutable.List.foreach(List.scala:381)
  at org.apache.spark.sql.execution.datasources.parquet.ParquetWriteSupport$.setSchema(ParquetWriteSupport.scala:446)
  at org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat.prepareWrite(ParquetFileFormat.scala:112)
  at org.apache.spark.sql.execution.datasources.FileFormatWriter$.write(FileFormatWriter.scala:134)
```